src/org/python/core/PyByteArray.java

```diff
@@ -926,32 +926,13 @@ public PyObject __add__(PyObject o) {
 
     @ExposedMethod(type = MethodType.BINARY, doc = BuiltinDocs.bytearray___add___doc)
     final synchronized PyObject bytearray___add__(PyObject o) {
-        PyByteArray sum = null;
+        // Duplicate this buffer
+        byte[] copy = new byte[size];
+        System.arraycopy(storage, offset, copy, 0, size);
+        PyByteArray sum = new PyByteArray(copy);
 
-        // XXX re-write using buffer API
-
-        if (o instanceof BaseBytes) {
-            BaseBytes ob = (BaseBytes)o;
-            // Quick route: allocate the right size bytearray and copy the two parts in.
-            sum = new PyByteArray(size + ob.size);
-            System.arraycopy(storage, offset, sum.storage, sum.offset, size);
-            System.arraycopy(ob.storage, ob.offset, sum.storage, sum.offset + size, ob.size);
-
-        } else if (o.getType() == PyString.TYPE) {
-            // Support bytes type, which in in Python 2.7 is an alias of str. Remove in 3.0
-            PyString os = (PyString)o;
-            // Allocate the right size bytearray and copy the two parts in.
-            sum = new PyByteArray(size + os.__len__());
-            System.arraycopy(storage, offset, sum.storage, sum.offset, size);
-            sum.setslice(size, sum.size, 1, os);
-
-        } else {
-            // Unsuitable type
-            // XXX note reversed order relative to __iadd__ may be wrong, matches Python 2.7
-            throw ConcatenationTypeError(TYPE, o.getType());
-        }
-
-        return sum;
+        // Concatenate the other buffer
+        return sum.__iadd__(o);
```

Comment:
I like that this is shorter and puts the logic in one place, but as you, not the avoidable copy that results when += inevitably finds there is no room for `o`. Everywhere else, this class does its utmost to avoid needless copying in the belief that sometimes byte arrays are big.

For this, it is only necessary to get the allocated size of the sum `bytearray` right, then it can look like the current handling of `PyString`.
```
            // Allocate the right size bytearray and copy the first part into it.
            sum = new PyByteArray(size + o.__len__());
            System.arraycopy(storage, offset, sum.storage, sum.offset, size);
            // Append o as bytes (if we can).
            return sum.bytearray___iadd__(o);
```
I'm seldom sure whether it is more appropriate to call e.g. `__iadd__` or `bytearray___iadd__`. Do we want a subclass to be able to intercept the action? I tend to think not here, as it is an implementation detail.

All types of `o` that may be added (and many others) have a `__len__` that won't throw, but I don't think we can check, other than by calling it. So we need a `try`. If you `catch` a `PyException`, you call `throw ConcatenationTypeError(TYPE, o.getType())`, and if it doesn't you now know how big the sum will be, unless the type is one `bytearray___iadd__` will root out.

Or maybe a quick check on the class of `o` is enough? I *think* all types of `o` that may be added (and a few others) bear the BufferProtocol, and have an `o.__len__()` it would be safe to call.

Reply:
Yes your right, I actually had the summed size at one point and changed it. I also agree its probably better to call `__bytearray___iadd__` directly, I will change that too.